### DISCUSSION TITLE
Mesh file for input

### DISCUSSION THREAD

**Discussion by hokonkim** [idaholab/moose#19097] [Q&A General] [answered]
Hello, I have a question about mesh file for input.

After running a moose, I can export the output file as *.e format, and

moose supports *.e file as a mesh(https://mooseframework.inl.gov/source/mesh/FileMesh.html).

What I am curious about is "if *.e file itself includes several results (e.g., t=0, result1, t=0.1, result2, t=0.2, result 3,...)",

how can I use a mesh at a specific time step?

(for example, I got a *.e file includes 8 results (time steps), but I want to use the mesh at time step 4) 

I appreciate your help.

Sincerely,
Hokon Kim

**Comment by ngrilli** [comment, marked as answer]
Dear @hokonkim 

If what you mean is that you want to use a "deformed" mesh from a mechanics simulation,
then the way to go is the restart option: 
https://mooseframework.inl.gov/application_usage/restart_recover.html

You can import the displacement variables from a previous simulation at the last time step
and then you run a new simulation based on that initial displacement/deformation.

You can see an example of restart a mechanics simulation in my github, using the SolutionUserObject:
https://github.com/ngrilli/c_pfor_am/blob/main/test/tests/DeleteElementsRestart/DummyVarRestart.i

Otherwise if you read the mesh from an .e file, you will always get the mesh from which the previous simulation started,
if you don't reinitialise the displacement,
I may be wrong in the case of mesh refinement.

If you are not talking about mechanics simulations,
then I am not sure what you mean.

Best Regards,
Nicolò

**Reply by hokonkim** [reply]
I was taking a look at what you suggested.

What I want to do is "using deformed mesh after applying postprocessor"

Specifically, at first input file..

    [Materials]
      [./tpbActivity]
        type  = ParsedMaterial
        block = 'PT_TPB_TET4'
        f_name = 'tpbActivity_S0'
        postprocessor_names = 'j_YSZ_bottom Poisoned'
        function = 'if(-j_YSZ_bottom > 1, if(aux_eta_tpb > Poisoned, 0, 4818), 4818)'
        args = 'aux_eta_tpb'
      [../]
    []    

After the first input file, I expected an output file which includes the deformed (postprocessor applied) mesh.

So, I made another input file to use this deformed mesh like below.

    [Mesh]
      [./file]
        type = FileMeshGenerator
        file = outputs/tpb_gan0108_2021-10-15-00:26-04.e
      [../]
    []

    [Materials]
      [./tpbActivity]
        type  = ParsedMaterial
        block = 'PT_TPB_TET4'
        f_name = 'tpbActivity_S0'
        function = '4818'
      [../]
    []

    [UserObjects]
      [./soln]
        type = SolutionUserObject
        mesh = outputs/tpb_gan0108_2021-10-15-00:26-04.e
        timestep = 'LATEST'
      [../]
    []

But, I'm stuck on how to initialize the "tpbActivity" from the previous output file.

How can I fix this problem?

Thanks.

**Reply by GiudGiud** [reply]
What is tpbActivity_S0 ? It's a function that is constant or comes from the previous solution?

**Reply by hokonkim** [reply]
tpbActivity(named as “tpbActivity_S0) is an assigned material property. 

Depending on a certain condition, it would be either activated or deactivated. 

Initializing material property from previous simulation is what I need, I think.

**Reply by GiudGiud** [reply]
ah my bad ParsedMaterial args are not clear. I ll rework that

You can use the Controls system to activate / deactivate blocks or control their value. 
https://mooseframework.inl.gov/syntax/Controls/index.html

You could use a SolutionFunction if the function can be initialized from a field/variable that is defined in the previous simulation
https://mooseframework.inl.gov/source/functions/SolutionFunction.html

**Reply by hokonkim** [reply]
Is there a way to store Material properties (ParsedMaterial) and use stored material properties as initial value?

I tried to store the material properties (MaterialRealAux), but it didn't work...

I think, if I can store the material property as variable (or aux_variable), then I could use SolutionUserObject...

Additionally...(part of input)

I tried to run an application as below.. but, it didn't work...

As far as I know, Output of Material properties is enabled by setting the "outputs" parameter.

Can I extract this value from output(*.e) file and import to another input file?

What will be the best solution for me?


    [Functions]
      [./aux_eta_tpb_ic]
        type = SolutionFunction
        solution = soln_aux_eta_tpb
      [../]
    []

    [Materials]
      [./tpbActivity]
        type  = ParsedMaterial
        block = 'PT_TPB_TET4'
        f_name = 'tpbActivity_S0'
        function = 'if(-j_YSZ_bottom > 1, if(aux_eta_tpb > Poisoned, 0, 4818), 4818)'
        args = 'aux_eta_tpb'
        postprocessor_names = 'j_YSZ_bottom Poisoned'
        outputs = exodus
      [../]
    []

    [UserObjects]
      [./soln_aux_eta_tpb]
        type = SolutionUserObject
        mesh = outputs/tpb_gan0108_2021-10-15-18:38-25.e
        timestep = 'LATEST'
        system_variables = aux_eta_tpb
      [../]
    []

    [AuxVariables]
      [./prev_aux_eta_tpb]
      [../]
    []

    [AuxKernels]
      [./soln_eta_tpb]
        type = SolutionAux
        solution = soln_aux_eta_tpb
        variable = prev_aux_eta_tpb
        direct = true
        execute_on = 'initial'
      [../]
    []

**Reply by GiudGiud** [reply]
You can do that., just use a MaterialRealAux to save the material property to an AuxVariable, then it will be saved to the solution file.

An other option, a little convoluted, is to save it to exodus (through an aux variable as well, but then you can use outputs='exodus' in the material property) then use that system to reload that into a variable, then into the material property

**Reply by hokonkim** [reply]
When I used MaterialRealAux like below

    [Functions]
      [./tpbActivity]
        type  = ParsedMaterial
        block = 'PT_TPB_TET4'
        f_name = 'tpbActivity_S0'
        function = 'if(-j_YSZ_bottom > 1, if(aux_eta_tpb > Poisoned, 0, 4818), 4818)'
        args = 'aux_eta_tpb'
        postprocessor_names = 'j_YSZ_bottom Poisoned'
        # Make values see at output(*.e) file
        outputs = exodus
      [../]
    []

    [AuxVariables]
      [./cp_tpbActivity_S0]
      [../]
    []

    [AuxKernels]
      [./copy_tpbActivity_S0]
        type = MaterialRealAux
        property = 'tpbActivity_S0'
        variable = cp_tpbActivity_S0
        execute_on = 'initial'
      [../]
    []

I got an error like this.

*** ERROR ***
The following error occurred in the object "copy_tpbActivity_S0", of type "MaterialRealAux".

Nodal AuxKernel 'copy_tpbActivity_S0' attempted to reference material property 'property'
Consider using an elemental auxiliary variable for 'cp_tpbActivity_S0'.




Also, when I tried another way (using exodus file), 'tpbActivity_S0' cannot be specificed..

I can see the 'tpbActivity_S0' in the output file, but I cannot use that name through SolutionUserObject.

Did I miss something?

**Reply by hokonkim** [reply]
Now, I got a previous "tpbActivity_S0" with the input below.

    [UserObjects]
      [./soln_tpbActivity_S0]
        type = SolutionUserObject
        mesh = outputs/tpb_gan0108_2021-10-15-18:38-25.e
        timestep = 'LATEST'
        system_variables = tpbActivity_S0
      [../]
    []

    [AuxVariables]
      [./prev_tpbActivity_S0]
        # Specifies the family of FE shape functions to use for this variable.
        # If the variable family is MONOMIAL then the AuxKernel will behave as an elemental.
        # Especially, Material property is "elemental", not "nodal".
        family = MONOMIAL
        order = CONSTANT
        block = 'PT_TPB_TET4'
      [../]
    []

    [AuxKernels]
      [./soln_tpbActivity]
        type = SolutionAux
        solution = soln_tpbActivity_S0
        variable = prev_tpbActivity_S0
        direct = true
        execute_on = 'initial'
      [../]
    []

I am still struggling to initialize material property with "prev_tpbActivity_S0".

I want to use the "prev_tpbActivity_S0" as initial material property condition, and

apply the function( 'if(-j_YSZ_bottom > 1, if(aux_eta_tpb > Poisoned, 0, 4818), 4818)' ) depending on time step.

I have no idea how to apply both (material initialization & apply function)....

because 

*** ERROR ***
Invalid function
if(time_check = 0, prev_tpbActivity_S0, if(-j_YSZ_bottom > 1, if(aux_eta_tpb > Poisoned, 0, 4818)))
aux_eta_tpb,prev_tpbActivity_S0,time_check,j_YSZ_bottom,Poisoned
in ParsedMaterialHelper.
Illegal number of parameters to function


Are there any tips for me?

Thanks!

**Reply by GiudGiud** [reply]
So for your first message, last part:
the solutionUO and the exodus are completely separate ways of restarting. When needed (and it should be avoidable most of the time) you can restart some fields using one and others using the other, but they act very separately. See more info here:
https://mooseframework.inl.gov/application_usage/restart_recover.html

For the second message,
can you paste your ParsedMaterial block here? To check the other arguments.
It seems one of your if statements is missing an argument, they should all have 3 arguments
`if(time_check = 0, prev_tpbActivity_S0, if(-j_YSZ_bottom > 1, if(aux_eta_tpb > Poisoned, 0, 4818), MISSING option))`

**Reply by hokonkim** [reply]
Here are some parts of input file.
- Use original mesh + "previous result from output(exodus) file"


      [Mesh]
        [./file]
          type = FileMeshGenerator
          file = outputs/original.inp
        [../]
      []

      [Materials]
        [./tpbActivity]
          type  = ParsedMaterial
          block = 'PT_TPB_TET4'
          f_name = 'tpbActivity_S0'
          # function = 'if(prev_tpbActivity_S0 > 0, if(-j_YSZ_bottom > 1, if(aux_eta_tpb > 0.163 * 0.92, 0, 4818) 4818), 0)'
          function = 'if(prev_tpbActivity_S0 > 0, if(aux_eta_tpb > Poisoned, 0, 4818), 0)'
          args = 'aux_eta_tpb prev_tpbActivity_S0'
          postprocessor_names = 'Poisoned'
          # Make values see at output(*.e) file
          outputs = exodus
        [../]
      []

      [UserObjects]
        [./soln_tpbActivity_S0]
          type = SolutionUserObject
          mesh = outputs/tpb_gan0108_2021-10-15-18:38-25.e
          timestep = 'LATEST'
          system_variables = tpbActivity_S0
        [../]
      []


      [AuxVariables]
        [./aux_eta_tpb]
          block = 'PT_TPB_TET4'
        [../]

        [./prev_tpbActivity_S0]
          # Specifies the family of FE shape functions to use for this variable.
          # If the variable family is MONOMIAL then the AuxKernel will behave as an elemental.
          # Especially, Material property is "elemental", not "nodal".
          family = MONOMIAL
          order = CONSTANT
          block = 'PT_TPB_TET4'
        [../]
      []

      [AuxKernels]
        [./eta_tpb]
          type = ParsedAux
          variable = aux_eta_tpb
          block = 'PT_TPB_TET4'
          function = '-R * T / 4 / F * log(7e-21 / p_O2) - (phi_LSM - phi_YSZ + 0.03607)'
          constant_names = 'R T F'
          constant_expressions = '8.3144598 1073 96485.33289' # (J/K/mol), (K), (C/mol)
          args = 'p_O2 phi_LSM phi_YSZ'
        [../]

        [./soln_tpbActivity]
          type = SolutionAux
          solution = soln_tpbActivity_S0
          variable = prev_tpbActivity_S0
          direct = true
          execute_on = 'initial'
        [../]
      []

      [Postprocessors]
        [./eta_total]
          type = FunctionValuePostprocessor
          function = 'funcOverpotential'
          outputs = 'console csv'
        [../]

        [./Poisoned]
          type = ParsedPostprocessor
          function = 'eta_total * 0 + poison * 0.92' # If you want to control the standard for poisoning, then change the number (0 ~ 1.0)
          pp_names = 'eta_total'
          constant_names = 'poison'
          constant_expressions = '0.163'
          outputs = 'console csv'
        [../]
      []

This was the best that I can do now.

Still, I wonder if the function in ParsedMaterial can use the function (commented out).

I tried.. but it didn't work (MOOSE didn't allow several if statements and several parameters..)

Are there ways to use several parameters or if-statement?

Thanks!

**Reply by GiudGiud** [reply]
Hello

- You are loading a .inp file, not an exodus file, so you cant use exodus restart. You'll have to work with SolutionUOs then
- This expression is still invalid ``    function = 'if(prev_tpbActivity_S0 > 0, if(-j_YSZ_bottom > 1, if(aux_eta_tpb > 0.163 * 0.92, 0, 4818) 4818), 0)'
``

it needs to be:
```    
function = 'if(prev_tpbActivity_S0 > 0, if(-j_YSZ_bottom > 1, if(aux_eta_tpb > 0.163 * 0.92, 0, 4818), 4818), 0)'
```
you are missing a comma

**Reply by hokonkim** [reply]
For now, I use above input file style to apply prev_tpbActivity_S0(from exodus) to Original mesh.

Through this way, I can run an application with previous(initialized) tpbActivity_S0 on Original mesh.

Is there more efficient way? 

For the function, I just knew I was missing a comma. So I tried again like below

I didn't get error for the function after adding comma.

      [./tpbActivity]
        type  = ParsedMaterial
        block = 'PT_TPB_TET4'
        f_name = 'tpbActivity_S0'
        function = 'if(prev_tpbActivity_S0 > 0, if(-j_YSZ_bottom > 1, if(aux_eta_tpb > Poisoned, 0, 4818), 4818), 0)'
        args = 'aux_eta_tpb prev_tpbActivity_S0'
        postprocessor_names = 'j_YSZ_bottom Poisoned'
        # Make values see at output(*.e) file
        outputs = exodus
      [../]

**Reply by GiudGiud** [reply]
prev_tpbActivity_S0 is a variable, so it can be restarted from a SolutionUO.
I think either that, or checkpoint restart, are the most efficient ways.
